src/utils/saml.ts

```diff
@@ -81,72 +134,73 @@ export const getPreValidateResponse = (
     "Response"
   );
 
+  const maybeIdpIssuer = getSamlIssuer(doc);
+
   const hasStrictValidation = pipe(
     O.fromNullable(strictValidationOptions),
-    O.chain(_ =>
+    O.chain(validationOptions =>
       pipe(
-        getSamlIssuer(doc),
-        O.chainNullableK(issuer => _[issuer])
+        maybeIdpIssuer,
+        O.chainNullableK(issuer => validationOptions[issuer])
       )
     ),
     O.getOrElse(() => false)
   );
 
-  interface IBaseOutput {
-    readonly InResponseTo: NonEmptyString;
-    readonly Assertion: Element;
-    readonly IssueInstant: Date;
-    readonly Response: Element;
-    readonly AssertionIssueInstant: Date;
-  }
-
-  interface ISamlCacheType {
-    readonly RequestXML: string;
-    readonly createdAt: Date;
-    readonly idpIssuer: string;
-  }
-
-  type IRequestAndResponseStep = IBaseOutput & {
-    readonly SAMLRequestCache: ISamlCacheType;
-  };
-
-  type ISAMLRequest = IRequestAndResponseStep & { readonly Request: Document };
+  const idpIssuer: string = pipe(
+    maybeIdpIssuer,
+    O.getOrElse(() => "not available")
+  );
 
-  type IIssueInstant = ISAMLRequest & {
-    readonly RequestIssueInstant: Date;
-    readonly RequestAuthnRequest: Element;
-  };
+  // here we are partially validating the response just to obtain a requestId (InResponseTo) before doing any more step.
+  // this is needed if we want to have to log the requestId at any step further
+  const errorOrPartiallyValidatedResponse: E.Either<
+    Error,
+    { readonly InResponseTo: NonEmptyString; readonly Response: Element }
+  > = pipe(
+    responsesCollection,
+    E.fromPredicate(
+      coll => coll.length < 2,
+      _ => new Error("SAML Response must have only one Response element")
+    ),
+    E.map(coll => coll.item(0)),
+    // this check is bound to the previous one, because we can receive no Response based on the official validator guidelines
+    E.chain(
+      E.fromNullable(new Error("Missing Response element inside SAML Response"))
+    ),
+    E.chain(Response =>
+      pipe(
+        NonEmptyString.decode(Response.getAttribute("InResponseTo")),
+        E.mapLeft(
+          () => new Error("InResponseTo must contain a non empty string")
+        ),
+        E.map(InResponseTo => ({ InResponseTo, Response }))
+      )
+    )
+  );
 
-  type IIssueInstantWithAuthnContextCR = IIssueInstant & {
-    readonly RequestAuthnContextClassRef: NonEmptyString;
-  };
+  const requestId: string = pipe(
+    errorOrPartiallyValidatedResponse,
+    E.map(({ InResponseTo }) => InResponseTo),
+    E.getOrElse(() => "not available")
```

Comment:
ditto

Reply:
fixed in https://github.com/pagopa/io-spid-commons/pull/121/commits/8a663a2b24774f521a9a08720d21d5a348d7a998